app/controllers/shopify_app/callback_controller.rb

```diff
@@ -68,15 +68,28 @@ def update_rails_cookie(api_session, cookie)
       ShopifyApp::Logger.debug("Saving Shopify user ID to cookie")
     end
 
-    def respond_successfully
+    def redirect_to_app
       if ShopifyAPI::Context.embedded?
-        return_to = session.delete(:return_to) || ""
-        redirect_to(ShopifyAPI::Auth.embedded_app_url(params[:host]) + return_to, allow_other_host: true)
+        return_to = "#{sanitized_host}#{session.delete(:return_to)}"
+        return_to = ShopifyApp.configuration.root_url if deduced_phishing_attack?
+        redirect_to(return_to, allow_other_host: true)
       else
         redirect_to(return_address)
       end
     end
 
+    def deduced_phishing_attack?
+      # host param doesn't match the configured myshopify_domain
+      sanitized_host.nil?
+    end
+
+    def sanitized_host
+      @sanitized_host ||= begin
+        decoded_host = ShopifyAPI::Auth.embedded_app_url(params[:host])
+        ShopifyApp::Utils.sanitize_shop_domain(decoded_host)
```

Comment:
Why are we adding this? This is what is causing the 404 on my end.
```
[5] pry(#<ShopifyApp::CallbackController>)> decoded_host = ShopifyAPI::Auth.embedded_app_url(params[:host])
=> "https://bill-klenotiz-testing-store.myshopify.com/admin/apps/1d0d4abc458c0d82d72d1849ffb01f63"
[6] pry(#<ShopifyApp::CallbackController>)> ShopifyApp::Utils.sanitize_shop_domain(decoded_host)
=> "bill-klenotiz-testing-store.myshopify.com"
```

Reply:
That's the security hole we are trying to patch :/ If `params[:host]` is to a hackerman's host, they can use trusted app domains to phish for information.